mne/preprocessing/faster.py

```diff
@@ -0,0 +1,463 @@
+            use_metrics.remove('eog_correlation')
+
+    bads = dict()
+    for m in use_metrics:
+        scores = np.atleast_2d(metrics[m](ica))
+        for s in scores:
+            b = find_outliers(s, thres, max_iter)
+            logger.info('Bad by %s:\n\t%s' % (m, b))
+            bads[m] = b
+
+    if return_by_metric:
+        return bads
+    else:
+        return np.unique(np.concatenate(list(bads.values()))).tolist()
+
+
+def faster_bad_channels_in_epochs(epochs, picks=None, thres=3,
+                                  use_metrics=None, max_iter=2,
+                                  return_by_metric=False):
+    """Implements the fourth step of the FASTER algorithm.
+
+    This function attempts to automatically mark bad channels in each epochs by
+    performing outlier detection.
+
+    Parameters
+    ----------
+    epochs : Instance of Epochs
+        The epochs to analyze.
+    picks : list of int | None
+        Channels to operate on. Defaults to EEG channels.
+    thres : float
+        The threshold value, in standard deviations, to apply. An epoch
+        crossing this threshold value is marked as bad. Defaults to 3.
+    use_metrics : list of str
+        List of metrics to use. Can be any combination of:
+            'amplitude', 'variance', 'deviation', 'median_gradient'
+        Defaults to all of them.
+    max_iter : int
+        The maximum number of iterations performed during outlier detection
+        (defaults to 2, original FASTER paper uses 1).
+    return_by_metric : bool
+        Whether to return the bad channels as a nested list (False, default) or
+        as a dictionary with the names of the used metrics as keys and the
+        bad channels found by this metric as a nested list.
+
+    Returns
+    -------
+    bads : list of lists of int
+        For each epoch, the indices of the bad channels.
+    """
+
+    metrics = {
+        'amplitude': lambda x: np.ptp(x, axis=2),
+        'deviation': lambda x: _deviation(x),
+        'variance': lambda x: np.var(x, axis=2),
+        'median_gradient': lambda x: np.median(np.abs(np.diff(x)), axis=2),
+        'line_noise': lambda x: _freqs_power(x, epochs.info['sfreq'],
+                                             [50, 60]),
+    }
+
+    if picks is None:
+        picks = pick_types(epochs.info, meg=False, eeg=True, exclude='bads')
+    if use_metrics is None:
+        use_metrics = metrics.keys()
+
+    data = epochs.get_data()
+
+    bads = defaultdict(lambda: [list() for _ in range(len(epochs))])
+    for ch_type, chs in _by_ch_type(epochs.info, picks):
+        for m in use_metrics:
+            logger.info('Bad channel-in-epoch detection on %s channels:'
+                        % ch_type.upper())
+            s_epochs = metrics[m](data[:, chs])
+            for i, s in enumerate(s_epochs):
+                b = [epochs.ch_names[picks[j]]
+                     for j in find_outliers(s, thres, max_iter)]
+                if len(b) > 0:
```

Comment:
same here ... single letter variable names. Notes for @dengemann if he's making a new PR